tests/integration/targets/keycloak_group/tasks/main.yml

```diff
@@ -0,0 +1,593 @@
+      - result_subsubsubgrp.end_state.realmRoles == []
+      - result_subsubsubgrp.end_state.subGroups == []
+
+## subgroup deletion tests
+## note: in principle we already have tested group deletion in general
+##   enough already, but what makes it interesting here again is to
+##   see it works also properly for subgroups and groups with subgroups
+- name: Deleting a subgroup by id (no parents needed)
+  community.general.keycloak_group:
+    auth_keycloak_url: "{{ url }}"
+    auth_realm: "{{ admin_realm }}"
+    auth_username: "{{ admin_user }}"
+    auth_password: "{{ admin_password }}"
+    realm: "{{ realm }}"
+    id: "{{ result_subsubsubgrp.end_state.id  }}"
+    state: absent
+  register: result
+
+- name: Debug
+  debug:
+    var: result
+
+- name: Assert that subgroup was deleted
+  assert:
+    that:
+      - result is changed
+      - result.end_state == {}
+
+- name: Redo subgroup deletion (idempotency test)
+  community.general.keycloak_group:
+    auth_keycloak_url: "{{ url }}"
+    auth_realm: "{{ admin_realm }}"
+    auth_username: "{{ admin_user }}"
+    auth_password: "{{ admin_password }}"
+    realm: "{{ realm }}"
+    id: "{{ result_subsubsubgrp.end_state.id  }}"
+    state: absent
+  register: result
+
+- name: Debug
+  debug:
+    var: result
+
+- name: Assert that nothing changed
+  assert:
+    that:
+      - result is not changed
+      - result.end_state == {}
+
+- name: Deleting a subgroup by name
+  community.general.keycloak_group:
+    auth_keycloak_url: "{{ url }}"
+    auth_realm: "{{ admin_realm }}"
+    auth_username: "{{ admin_user }}"
+    auth_password: "{{ admin_password }}"
+    realm: "{{ realm }}"
+    name: new-subgrp1
+    parents:
+      - rootgrp
+    state: absent
+  register: result
+
+- name: Debug
+  debug:
+    var: result
+
+- name: Assert that subgroup was deleted
+  assert:
+    that:
+      - result is changed
+      - result.end_state == {}
+
+- name: Redo deleting a subgroup by name (idempotency test)
+  community.general.keycloak_group:
+    auth_keycloak_url: "{{ url }}"
+    auth_realm: "{{ admin_realm }}"
+    auth_username: "{{ admin_user }}"
+    auth_password: "{{ admin_password }}"
+    realm: "{{ realm }}"
+    name: new-subgrp1
+    parents:
+      - rootgrp
+    state: absent
+  register: result
+
+- name: Debug
+  debug:
+    var: result
+
+- name: Assert that nothing has changed
+  assert:
+    that:
+      - result is not changed
+      - result.end_state == {}
+
+- name: Delete keycloak group which has subgroups
+  community.general.keycloak_group:
+    auth_keycloak_url: "{{ url }}"
+    auth_realm: "{{ admin_realm }}"
+    auth_username: "{{ admin_user }}"
+    auth_password: "{{ admin_password }}"
+    realm: "{{ realm }}"
+    name: rootgrp
+    state: absent
+  register: result
+
+- name: Debug
+  debug:
+    var: result
+
+- name: Assert that group was deleted
+  assert:
+    that:
+      - result is changed
+      - result.end_state == {}
```

Comment:
What happens to the subgroups in this case? Do they become top level automatically or is the deletion cascaded to them? Should we want to have both options controlled by flag?

Reply:
Yeah, deleting a group automatically also deletes its subgroups. This seems to be general keycloak behaviour as the same happens when you delete a group using the UI. Technically it would be possible to handle things differently in the ansible module, but I don't know.

Maybe one could do a variant of your suggestion. Checking on deletion if a group has sub-groups and failing on default for this case with an optional `force` or `recursive` flag similar to how `rm` handles directories, as a safety feature.

Reply:
OK, despite this being keycloak's default behaviour, I think it would be for the best to document that explicitly for the benefit of unwary users. I agree that there is no point in the module behaving differently.

OTOH, another question that occurred to me: is it possible to remove the parent-child relationship? (That sounds eerily like kidnapping, I know). Let's say we have group P with children A, B, C. User wants to remove P and children except for group B. Is it possible to remove B from under P?

Or maybe another to put it would be: what happens if we run the module twice (like in the idempotency tests you wrote), but in the second run we do not specify `parents`? Will the parent be "removed from the group"? Should it?

About a force option, I think that would fall outside the intent of this PR, so maybe you might want to create a new one for that.

Reply:
> OK, despite this being keycloak's default behaviour, I think it would be for the best to document that explicitly for the benefit of unwary users. I agree that there is no point in the module behaving differently.

Yeah, being more explicit in the docu cannot hurt I guess. Changed the description of the module state parameter like this:

````
- On C(absent), the group will be removed if it exists. Be aware that absenting a group with subgroups will automatically delete all its subgroups too.
````

Reply:
> About a force option, I think that would fall outside the intent of this PR, so maybe you might want to create a new one for that.

Agreed, especially as it would be at least in theory a breaking change. Admittedly it might be a bit abscure, but image a situation where an user created a playbook which might absent a group and this group happens to have some subgroups by other means (manually added, directly using rest-api with ansible uri module or similar, ...). No problem with current version of the module which did not handle subgroups at all but with the proposed opt-out `force` flag it would start to fail.

Reply:
> Or maybe another to put it would be: what happens if we run the module twice (like in the idempotency tests you wrote), but in the second run we do not specify `parents`? Will the parent be "removed from the group"? Should it?

No, current proposed implementation would not change anything for a subgroup independend if `parents` is set or not. One can thing of `parents` parameter more of a helper value to properly "locate" the group in question instead of really carrying some "state" of the group. In that sense what exactly happens depends again if the user is working with group names or ID's.

For the 1st case ansible would actually do its work against a totally different group, one which happens to have the same name as the subgroup of the first module call but as no parents are defined must be top level (base group).

With ID's any existing keycloak group can be directly accessed from the keycloak api, doesn't matter if it is a base or a subgroup. So when using its ID one always gets the right group and specifying parents is actually not necessary.

> OTOH, another question that occurred to me: is it possible to remove the parent-child relationship? (That sounds eerily like kidnapping, I know). Let's say we have group P with children A, B, C. User wants to remove P and children except for group B. Is it possible to remove B from under P?

Well this sounds for me actually like a different use case maybe even having its own `state` value (`move`, `migrate`, ...?) and in the general case one would actually need to specify two sets of parents (start and endpoint) with each one possibly being unset if start or endpoint is the top level.